File: dotnet/Gitologist/Branch.cs
using System.Text.RegularExpressions;

namespace Gitologist;

internal static class Branch
{
    internal static async Task<string> GetCurrentBranch(string gitDir)
    {
        var headPath = Path.Combine(gitDir, "HEAD");
        var headContent = (await File.ReadAllTextAsync(headPath)).Trim();

        var match = Regex.Match(headContent, @"^ref: refs\/heads\/(.+)$");
        if (match.Success)
        {
            return match.Groups[1].Value;
        }

        throw new InvalidOperationException("Not on a branch (detached HEAD)");
    }

    internal static async Task<string?> GetCurrentCommit(string gitDir)
    {
        try
        {
            var branch = await GetCurrentBranch(gitDir);
            var branchPath = Path.Combine(gitDir, "refs", "heads", branch);

            if (!File.Exists(branchPath))
            {
                return null;
            }

            return (await File.ReadAllTextAsync(branchPath)).Trim();
        }
        catch
        {
            return null;
        }
    }
}

File: web/src/branch.ts
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { join } from "node:path";

export async function getCurrentBranch(gitDir: string): Promise<string> {
	const headPath = join(gitDir, "HEAD");
	const headContent = (await readFile(headPath, "utf-8")).trim();

	const match = headContent.match(/^ref: refs\/heads\/(.+)$/);
	if (match) {
		return match[1];
	}

	throw new Error("Not on a branch (detached HEAD)");
}

export async function getCurrentCommit(gitDir: string): Promise<string | null> {
	try {
		const branch = await getCurrentBranch(gitDir);
		const branchPath = join(gitDir, "refs", "heads", branch);

		if (!existsSync(branchPath)) {
			return null;
		}

		return (await readFile(branchPath, "utf-8")).trim();
	} catch {
		return null;
	}
}
